pconn/connmanager.go

```diff
@@ -0,0 +1,445 @@
+
+	// defaultRetryDuration is the default duration of time for retrying
+	// persistent connections.
+	defaultRetryDuration = time.Second * 5
+)
+
+// ConnState represents the state of the requested connection.
+type ConnState uint8
+
+// ConnState can be either pending, established, disconnected or failed.  When
+// a new connection is requested, it is attempted and categorized as
+// established or failed depending on the connection result.  An established
+// connection which was disconnected is categorized as disconnected.
+const (
+	ConnPending ConnState = iota
+	ConnEstablished
+	ConnDisconnected
+	ConnFailed
+)
+
+// Req is the connection request to a network address. All requests are assumed
+// to be permanent, and will be retried until success or cancellation.
+type Req struct {
+	// The following variables must only be used atomically.
+	id uint64
+
+	Addr net.Addr
+
+	conn       net.Conn
+	state      ConnState
+	stateMtx   sync.RWMutex
+	retryCount uint32
+}
+
+// updateState updates the state of the connection request.
+func (c *Req) updateState(state ConnState) {
+	c.stateMtx.Lock()
+	c.state = state
+	c.stateMtx.Unlock()
+}
+
+// ID returns a unique identifier for the connection request.
+func (c *Req) ID() uint64 {
+	return atomic.LoadUint64(&c.id)
+}
+
+// State is the connection state of the requested connection.
+func (c *Req) State() ConnState {
+	c.stateMtx.RLock()
+	state := c.state
+	c.stateMtx.RUnlock()
+	return state
+}
+
+// String returns a human-readable string for the connection request.
+func (c *Req) String() string {
+	if c.Addr.String() == "" {
+		return fmt.Sprintf("reqid %d", atomic.LoadUint64(&c.id))
+	}
+	return fmt.Sprintf("%s (reqid %d)", c.Addr, atomic.LoadUint64(&c.id))
+}
+
+// Config holds the configuration options related to the connection manager.
+type Config struct {
+	// Listeners defines a slice of listeners for which the connection
+	// manager will take ownership of and accept connections.  When a
+	// connection is accepted, the OnAccept handler will be invoked with the
+	// connection.  Since the connection manager takes ownership of these
+	// listeners, they will be closed when the connection manager is
+	// stopped.
+	//
+	// This field will not have any effect if the OnAccept field is not
+	// also specified.  It may be nil if the caller does not wish to listen
+	// for incoming connections.
+	Listeners []net.Listener
+
+	// OnAccept is a callback that is fired when an inbound connection is
+	// accepted.  It is the caller's responsibility to close the connection.
+	// Failure to close the connection will result in the connection manager
+	// believing the connection is still active and thus have undesirable
+	// side effects such as still counting toward maximum connection limits.
+	//
+	// This field will not have any effect if the Listeners field is not
+	// also specified since there couldn't possibly be any accepted
+	// connections in that case.
+	OnAccept func(net.Conn)
```

Comment:
On second thought, perhaps we can just apply the OG patch (the one I started, but you fixed) to my fork, which will then eventually be submitted back upstream to `btcd`.  This way, it will actually also be possible to review the added diff. As currently presented, it's difficult to see exactly how the original `connmgr` package was modified.

Reply:
It would also significantly shrink the size of this diff!

Reply:
Bulk of the PR has been moved to https://github.com/Roasbeef/btcd/pull/10, thinking this PR will just become the glide update/minor fixes?